src/choleskyQR.jl

```diff
@@ -51,143 +51,200 @@ function cholesky_jacobimatrix(W::AbstractMatrix, Q)
+    U = F.R
+    U = ApplyArray(*,Diagonal(sign.(view(U,band(0)))),U)  # QR decomposition does not force positive diagonals on R by default
     X = jacobimatrix(P)
     UX = ApplyArray(*,U,X)
-    ev = zeros(T,2) # compute a length 2 vector on first go
-    ev[1] = dot(view(UX,1,1:2), U[1:2,1:2] \ [zero(T); one(T)])
-    ev[2] = dot(view(UX,2,1:3), U[1:3,1:3] \ [zeros(T,2); one(T)])
-    return QRJacobiBand{:ev,T}(ev, U, UX, 2)
+    dv = Vector{T}(undef,2) # compute a length 2 vector on first go
+    ev = Vector{T}(undef,2)
+    dv[1] = UX[1,1]/U[1,1] # this is dot(view(UX,1,1), U[1,1] \ [one(T)])
+    dv[2] = -U[1,2]*UX[2,1]/(U[1,1]*U[2,2])+UX[2,2]/U[2,2] # this is dot(view(UX,2,1:2), U[1:2,1:2] \ [zero(T); one(T)])
+    ev[1] = -UX[1,1]*U[1,2]/(U[1,1]*U[2,2])+UX[1,2]/U[2,2] # this is dot(view(UX,1,1:2), U[1:2,1:2] \ [zero(T); one(T)])
+    ev[2] = UX[2,1]/U[3,3]*(-U[1,3]/U[1,1]+U[1,2]*U[2,3]/(U[1,1]*U[2,2]))+UX[2,2]/U[3,3]*(-U[2,3]/U[2,2])+UX[2,3]/U[3,3] # this is dot(view(UX,2,1:3), U[1:3,1:3] \ [zeros(T,2); one(T)])
+    return QRJacobiData{:R,T}(dv, ev, U, UX, P, 2)
 end
 
-size(::QRJacobiBand) = (ℵ₀,) # Stored as an infinite cached vector
+
+size(::QRJacobiData) = (ℵ₀,2) # Stored as two infinite cached bands
+
+function getindex(K::QRJacobiData, n::Integer, m::Integer)
+    @assert (m==1) || (m==2)
+    resizedata!(K,n,m)
+    m == 1 && return K.dv[n]
+    m == 2 && return K.ev[n]
+end
 
 # Resize and filling functions for cached implementation
-function resizedata!(K::QRJacobiBand, nm::Integer)
+function resizedata!(K::QRJacobiData, n::Integer, m::Integer)
+    nm = max(n,m)
     νμ = K.datasize
     if nm > νμ
-        resize!(K.data,nm)
-        cache_filldata!(K, νμ:nm)
+        resize!(K.dv,nm)
+        resize!(K.ev,nm)
+        _fillqrbanddata!(K, νμ:nm)
         K.datasize = nm
     end
     K
 end
-function cache_filldata!(J::QRJacobiBand{:dv,T}, inds::UnitRange{Int}) where T
-    # pre-fill U and UX to prevent expensive step-by-step filling in of cached U and UX in the loop
-    getindex(J.U,inds[end]+1,inds[end]+1)
-    getindex(J.UX,inds[end]+1,inds[end]+1)
-
-    ek = [zero(T); one(T)]
-    @inbounds for k in inds
-        J.data[k] = dot(view(J.UX,k,k-1:k), J.U[k-1:k,k-1:k] \ ek)
+function _fillqrbanddata!(J::QRJacobiData{:Q,T}, inds::UnitRange{Int}) where T
+    b = bandwidths(J.U.factors)[2]÷2
+    # pre-fill cached arrays to avoid excessive cost from expansion in loop
+    m, jj = 1+inds[end], inds[2:end]
+    X = jacobimatrix(J.P)[1:m+b+2,1:m+b+2]
+    resizedata!(J.U.factors,m+b,m+b)
+    resizedata!(J.U.τ,m)
+    K, τ, F, dv, ev = J.UX, J.U.τ, J.U.factors, J.dv, J.ev
+    D = sign.(view(J.U.R,band(0)).*view(J.U.R,band(0))[2:end])
+    w = Vector{T}(undef,b+3)
+    M = Matrix{T}(undef,b+3,b+3)
+    @inbounds for n in jj
+        dv[n] = K[1,1] # no sign correction needed on diagonal entry due to cancellation
+        doublehouseholderapply!(K,τ[n+1],[zero(T);one(T);F[n+2:n+b+2,n+1]],w)
```

Comment:
This is still allocating. Probably change `doublehoseholderapply!` to know that the first two entries of `v` are zero and 1.

Reply:
so it's passing the vcat vector here that is allocating?

Reply:
this allocates twice: first by calling `F[n+2:n+b+2,n+1]` and second by concatenating a 0 and a 1

Reply:
ok I guess the first of these can be eliminated by passing a view instead. Then the function can assume, except for the n=1 special case, that the entries are 0 and 1.

Reply:
passing the view seems to be better but I have a feeling `mul!(w, M, [zero(T);one(T);v])` would still allocate? What's a sensible way around that which isn't rewriting matrix multiplication?

Reply:
Probably just use `reflectorApply!`?

https://github.com/JuliaLang/julia/blob/3051f47a40c37a2e898b9814856f7d4173b6d8c4/stdlib/LinearAlgebra/src/generic.jl#L1595

Reply:
I have experimented with modifications of that, can't seem to get it to reduce allocations using that. I will keep trying for now.

That said, having benchmarked a lot of different variations, the current variant pushes allocations under the current `R` method implementation so that's nice.

```julia
julia> @time Rtimer(100000,f,P,sqrtW)
  0.618029 seconds (11.78 M allocations: 526.182 MiB, 22.24% gc time)
0.617991553

julia> @time Qtimer(100000,f,P,sqrtW)
  0.654008 seconds (9.39 M allocations: 643.110 MiB, 19.61% gc time)
0.653952685
```

Reply:
Did you look at `reflectorApply!`? You aren't using the QR data structure in the right way, you should be following existing coding patterns. (And of course the code is still allocating if you are calling `[zero(T); one(T); v]`.)

Note there is no need to concatenate a `1` because it just ignores the first entry. Eg the following code
```julia
    v = [one(T);view(F.factors,2:b,1)]
     mul!(w, M, v)  # M is symmetric
     M .= M .- F.τ[1] .* v .* w'
     mul!(w, M, v)
     M .= M .- F.τ[1] .* w .* v'
```
should just be written
```
    v = view(F.factors,1:b,1)
    reflectorApply!(v, F.τ[1], M)
    reflectorApply!(v, F.τ[1], M')
```
And the calls to `doublehouseholderapply!` similarly modified similarly, that is DO NOT concatenate a 0 and a 1 but instead call `reflectorApply!` with the right views of each piece.

Reply:
Note this also means we don't even need the working variable `w`

Reply:
I tried something similar to that but ended up with more allocations, let's see if this template can get me there

Reply:
I think the part I may have been missing is that I can ignore the first entry, it may work like this

Reply:
I have to say I don't get `reflectorApply!(v, τ, M)`. I would have thought all it does is in-place left apply `I - τ*v*v'` to `M` and return this by overwriting `M` but evidently that is not what it does since using it in this way in the general case gives completely wrong entries.

Reply:
It does `(I - τ*[1;v]*[1;v]')*M` in-place.

Reply:
Ah, I see. Then maybe I should first rewrite it into that style in my implementation, then it should be easy to just swap this function in.